entity.go

```diff
@@ -605,22 +651,52 @@ func deserializeStruct(dst interface{}, b []byte) error {
 	fieldMap, _ := getFieldInfoAndMetadata(t)
 	var notFoundField string
 
-	for _, metaData := range smd.metaDatas {
-		fieldName, slice, zeroValue := metaData, false, false
-		if metaData[0] == '$' {
-			fieldName, slice = metaData[1:], true
-		} else if metaData[0] == '!' {
-			fieldName, zeroValue = metaData[1:], true
+	pls, hasPLS := dst.(datastore.PropertyLoadSaver)
+	switch header {
+	case serializationStateNormal:
+		if hasPLS {
+			// unfortunately fetching PropertyList cache to non-PLS is not supported
+			// and should falls back to datastore
+			return errCacheFetchFailed
 		}
-		nameParts := strings.Split(fieldName, ".")
+		for _, metaData := range smd.metaDatas {
+			fieldName, slice, zeroValue := metaData, false, false
+			if metaData[0] == '$' {
+				fieldName, slice = metaData[1:], true
+			} else if metaData[0] == '!' {
+				fieldName, zeroValue = metaData[1:], true
+			}
+			nameParts := strings.Split(fieldName, ".")
 
-		if fi, ok := fieldMap[fieldName]; ok {
-			if err := deserializeStructInternal(sd.dec, fi, fieldName, nameParts, slice, zeroValue, structHistory, v, t); err != nil {
+			if fi, ok := fieldMap[fieldName]; ok {
+				if err := deserializeStructInternal(sd.dec, fi, fieldName, nameParts, slice, zeroValue, structHistory, v, t); err != nil {
+					return err
+				}
+			} else {
+				sd.dec.Decode(nil) // Discard the value
+				notFoundField = fieldName
+			}
+		}
+	case serializationStatePropertyList:
+		var props []datastore.Property
+		if err := sd.dec.Decode(&props); err != nil {
+			return errCacheFetchFailed
+		}
+		for i := 0; i < len(props); i++ {
+			nilVal := nilType{}
+			if props[i].Value == nilVal {
```

Comment:
Compare to the new `nilValue` instead of the local `nilVal`.

Reply:
Fixed 0ffa7caaada1ace69f367ee36d201bcebed76bbd